File: source/com/barrybecker4/puzzle/shuffle1/Deck2.java
package com.barrybecker4.puzzle.shuffle1;

import com.barrybecker4.common.math.MathUtil;

/**
 * Analytic solution to the card shuffling problem.
 * Web references:
 *  http://www.math.ucsd.edu/~ronspubs/83_05_shuffles.pdf
 *  http://www.oreillynet.com/pub/wlg/5094 (Scott's original link, now broken)
 *  https://www.math.hmc.edu/funfacts/ffiles/20001.1-6.shtml
 *  https://coderanch.com/t/35113/Card-Shuffle
 *
 * @author Scott Sauyet
 */
public class Deck2 implements Deck {

    private int[] cards;
    private int count;

    public long shuffleUntilSorted(int iCut) {

        //Deck2 deck = new Deck2(nCards);
        this.doPerfectShuffle(iCut);
        return MathUtil.lcm(getCycles());
    }

    Deck2(int count) {

        if (count < 1) {
            throw new IllegalArgumentException("Deck must contain at least " +
                                    "one card.  You entered " + count + '.');
        }

        this.count = count;
        cards = new int[count];
        for (int i = 0 ; i < count; i++) {
            cards[i] = i;
        }
    }

    public int size() {
        return cards.length;
    }

    public int get(int i) {
        return cards[i];
    }

    public void doPerfectShuffle(int cut) {

        cut %= count;
        int[] top = new int[cut];

        System.arraycopy(cards, 0, top, 0, cut);

        int[] bottom = new int[count - cut];

        System.arraycopy(cards, cut, bottom, 0, count - cut);

        int[] updated = new int[count];
        int shared = Math.min(cut, count - cut);
        int different = count - shared - shared;
        int[] extras = new int[different];

        if (cut > count - cut) {
            System.arraycopy(top, 0, extras, 0, different);
        } else {
            System.arraycopy(bottom, 0, extras, 0, different);
        }

        for (int i = 0; i < shared; i++) {
            updated[count - (2 * i) - 1] = top[cut - i - 1];
            updated[count - (2 * i) - 2] = bottom[count - cut - i - 1];
        }

        System.arraycopy(extras, 0, updated, 0, different);
        cards = updated;
    }

    private int[] getCycles() {

        int[] cycles = new int[count];

        for (int i = 0; i < count; i++) {

            int index = i;
            int result = 1;
            while (cards[index] != i) {
                result++;
                index = cards[index];
            }
            cycles[i] = result;
        }

        return cycles;
    }

    @Override
    public String toString() {

        StringBuilder buffer = new StringBuilder();
        buffer.append("Deck [");
        buffer.append(cards[0]);

        for (int i = 1; i < count; i++) {
            buffer.append(", ");
            buffer.append(cards[i]);
        }

        buffer.append("]");
        return buffer.toString();
    }


    public static void main(String[] args) {

        assert (args.length == 2) : "Usage: java Deck nCards iCut";

        int cards = 1002;
        int cut = 101;

        try {
            cards = Integer.parseInt(args[0]);
            cut = Integer.parseInt(args[1]);
        } catch (NumberFormatException nfe) {
            System.out.println("Arguments must be numeric.");
            System.exit(2);
        }

        long start = System.currentTimeMillis();
        Deck2 deck = new Deck2(cards);
        long result = deck.shuffleUntilSorted(cut);
        long time = System.currentTimeMillis() - start;

        System.out.println("A perfect shuffleUntilSorted on " + cards + " cards, cut " + cut
                           + " deep, takes " + result + " iterations to restore"
                           + " the deck.");

        System.out.println("Calculation performed in " + time + "ms.");
    }
}

File: scala-source/com/barrybecker4/puzzle/shuffle/Deck2.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.shuffle

import com.barrybecker4.math.MathUtil

/**
  * Analytic solution to the card shuffling problem (order of the permutation = lcm of cycle lengths).
  * Web references:
  * http://www.math.ucsd.edu/~ronspubs/83_05_shuffles.pdf
  * http://www.oreillynet.com/pub/wlg/5094 (Scott's original link, now broken)
  * https://www.math.hmc.edu/funfacts/ffiles/20001.1-6.shtml
  * https://coderanch.com/t/35113/Card-Shuffle
  * @author Scott Sauyet
  */
object Deck2 {

  def main(args: Array[String]): Unit = {
    val (cards, cut) = args.length match {
      case 0 =>
        // Defaults when run from Gradle/IDE without program arguments (matches original demo).
        (1002, 101)
      case 2 =>
        try (args(0).toInt, args(1).toInt)
        catch {
          case _: NumberFormatException =>
            println("Arguments must be numeric.")
            sys.exit(2)
        }
      case _ =>
        println("Usage: Deck2 [nCards iCut]")
        println("  With no arguments, uses 1002 cards and cut 101.")
        sys.exit(1)
    }
    val start = System.currentTimeMillis()
    val deck = new Deck2(cards)
    val result = deck.shuffleUntilSorted(cut)
    val time = System.currentTimeMillis() - start
    println(
      "A perfect shuffle on " + cards + " cards, cut " + cut
        + " deep, takes " + result + " iterations to restore the deck."
    )
    println("Calculation performed in " + time + "ms.")
  }
}

class Deck2(val count: Int) extends Deck {
  if (count < 1)
    throw new IllegalArgumentException("Deck must contain at least one card. You entered " + count + '.')

  private var cards: Array[Int] = Array.tabulate(count)(identity)

  def shuffleUntilSorted(iCut: Int): Long = {
    doPerfectShuffle(iCut)
    MathUtil.lcm(cycleLengthsForPermutation)
  }

  def size: Int = cards.length

  def get(i: Int): Int = cards(i)

  def doPerfectShuffle(theCut: Int): Unit =
    PerfectShuffle.applyInPlace(cards, theCut)

  /** Length of the cycle containing `i` under `j -> cards(j)` (matches the original counting scheme). */
  private def cycleLengthFrom(i: Int): Int = {
    var index = i
    var result = 1
    while (cards(index) != i) {
      result += 1
      index = cards(index)
    }
    result
  }

  private def cycleLengthsForPermutation: Array[Int] = {
    val cycles = new Array[Int](count)
    for (i <- 0 until count) cycles(i) = cycleLengthFrom(i)
    cycles
  }

  override def toString: String = "Deck [" + cards.mkString(", ") + "]"
}
